t/proxyunits.t

```diff
@@ -771,46 +784,138 @@ check_sanity($ps);
         );
     };
 
-    my $cmd;
-    my $key;
+    subtest 'await(r, p, 1): send [h], recv 1 response and 2 reconn' => sub {
+        my $key = "/awaitone/a";
+        my $ps_send = "get $key\r\n";
+        proxy_test(
+            ps_send => $ps_send,
+            be_recv => {0 => [$ps_send], 1 => [$ps_send], 2 => [$ps_send]},
+            be_send => {0 => ["VALUE $key 0 2\r\nok\r\nEND\r\n"]},
+            ps_recv => ["VALUE $key 0 1\r\n", "1\r\n", "END\r\n"],
+        );
+        # reconnect due to timeout
+        $mbe[1] = accept_backend($mocksrvs[1]);
+        $mbe[2] = accept_backend($mocksrvs[2]);
+    };
 
-    # await(r, p, 1)
-    $key = "/awaitone/a";
-    $cmd = "get $key\r\n";
-    print $ps $cmd;
-    for my $be (@mbe) {
-        is(scalar <$be>, $cmd, "awaitone backend req");
-        print $be "VALUE $key 0 2\r\nok\r\nEND\r\n";
-    }
-    is(scalar <$ps>, "VALUE $key 0 1\r\n", "response from await");
-    is(scalar <$ps>, "1\r\n", "looking for a single response");
-    is(scalar <$ps>, "END\r\n", "end from await");
+    subtest 'await(r, p, 1): send [h, m, m], recv 1 response' => sub {
+        my $key = "/awaitone/a";
+        my $ps_send = "get $key\r\n";
+        my $be_hit = "VALUE $key 0 2\r\nok\r\nEND\r\n";
+        my $be_miss = "END\r\n";
+        proxy_test(
+            ps_send => $ps_send,
+            be_recv => {0 => [$ps_send], 1 => [$ps_send], 2 => [$ps_send]},
+            be_send => {0 => [$be_hit]},
+            ps_recv => ["VALUE $key 0 1\r\n", "1\r\n", "END\r\n"],
+        );
+        # send response to not timeout
+        $mbe[1]->send($be_miss);
+        $mbe[2]->send($be_miss);
+    };
 
-    # await(r, p(3+), 2)
-    $key = "/awaitone/b";
```

Comment:
It's hard to see because the tests are so verbose now, but I don't think I see this one replaced below?

this was testing 3 pools wait for 2 responses?

Reply:
look for

`subtest 'await(r, p, 2): sent [h, h, h], recv 2 responses' => sub {`

Reply:
I am hoping the title of the subtest make some sense in determining the test behaviors.

Reply:
the title is a goal, not an execution plan.

Reply:
Happy to change to good alternatives.

Reply:
I'm not really sure how to explain it yet, I'm sorry.

In the old test format I could see a comment (the plan/goal) and then the steps linearly.

When I'm reviewing the tests now, the steps are:
- look at the goal
- look at the argument list, determine if it matches the goal
- look at the function and execute it in my head to see if the execution matches the plan

I'm just stupid/slow. sorry.

Reply:
Your steps makes sense, and I was trying to serve those steps in my way:

Let me explain :) 
`subtest 'await(r, p, 2): sent [h, h, h], recv 2 responses' => sub {` in which:

`recv 2 responses` is the goal
`sent [h, h, h],` is the argument list or input
`await(r, p, 2)` is the function

It can be in any other form/format as long as it serves the purpose, sometimes it is hard for me to get it right without knowing exactly what you are suggesting (not complaint at all, communication is just hard.).

Reply:
If you would like me to convert them into comment, I am happy to do so.

Reply:
I think my point is that as a reviewer I need to make sure the test code matches the comment. I can read the note just fine, moving it to a comment doesn't change the functionality of the sentence :)

I also need to make sure the test is exercising the memcached code in a way that is useful, so there's possibly more going on in my head than an outsider.

Reply:
> moving it to a comment doesn't change the functionality of the sentence :)

That's what I thought :)

Reply:
Understood, I am doing what I can to make that easy for you as the reviewer.

Reply:
Actually I think you misread my original breakdown:

- look at the goal: `await(r, p, 2): sent [h, h, h], recv 2 responses`
- look at the argument list: `my $blah; proxy_test(etc, etc, etc, etc)`
- look at the function and execute it in my head: `sub proxy_test { arg arg loop test loop test }`

previously this was all linear in place, and if I wanted to adjust or add debugging to any step it was also in-place and independent of any other test being executed.

Reply:
`sub proxy_test { arg arg loop test loop test }` is generic, and soon you wouldn't need to look into `proxy_test`